Holovibes/cuda_sources/bw_area_filter.cu

```diff
@@ -0,0 +1,338 @@
+                                                       height,
+                                                       size_t_gpu_);
+    first_pass_kernel4<<<blocks, threads, 0, stream>>>(image_d,
+                                                       labels_d,
+                                                       linked_d,
+                                                       labels_sizes_d,
+                                                       width,
+                                                       height,
+                                                       size_t_gpu_);
+}
+
+__global__ void second_pass_kernel(size_t* labels_d, size_t size, size_t* linked_d, size_t* labels_sizes_d)
+{
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (idx < size && labels_d[idx] != 0)
+    {
+        size_t l = labels_d[idx];
+        get_linked_label(&l, linked_d);
+        labels_d[idx] = l;
+        atomicAdd(labels_sizes_d + l, 1);
+    }
+}
+
+void get_connected_component(const float* image_d,
+                             size_t* labels_d,
+                             size_t* labels_sizes_d,
+                             size_t* linked_d,
+                             size_t* size_t_gpu_,
+                             const size_t width,
+                             const size_t height,
+                             const cudaStream_t stream)
+{
+    size_t size = width * height;
+    uint threads = get_max_threads_1d();
+    uint blocks = map_blocks_to_problem(size, threads);
+
+    cudaXMemset(labels_d, 0, size * sizeof(size_t));
+    cudaXMemset(labels_sizes_d, 0, size * sizeof(size_t));
+
+    first_pass(image_d, labels_d, linked_d, labels_sizes_d, size_t_gpu_, width, height, stream);
+
+    second_pass_kernel<<<blocks, threads, 0, stream>>>(labels_d, size, linked_d, labels_sizes_d);
+}
+
+__device__ void swap(size_t* T, size_t i, size_t j)
+{
+    size_t tmp = T[i];
+    T[i] = T[j];
+    T[j] = tmp;
+}
+
+__global__ void _get_n_max_index(size_t* labels_size_d, size_t nb_label, size_t* labels_max_d, size_t n)
+{
+    for (size_t i = 0; i < n; i++)
+    {
+        size_t j = i;
+        labels_max_d[j] = j;
+        while (j > 0 && labels_size_d[labels_max_d[j - 1]] > labels_size_d[labels_max_d[j]])
+        {
+            swap(labels_max_d, j, j - 1);
+            j--;
+        }
+    }
+    for (size_t i = n; i < nb_label; i++)
+    {
+        if (labels_size_d[i] > labels_size_d[labels_max_d[0]])
+        {
+            labels_max_d[0] = i;
+            size_t j = 1;
+            while (j < n && labels_size_d[labels_max_d[j - 1]] > labels_size_d[labels_max_d[j]])
+            {
+                swap(labels_max_d, j, j - 1);
+                j++;
+            }
+        }
+    }
+}
+
+void get_n_max_index(size_t* labels_sizes_d, size_t nb_label, size_t* labels_max_d, size_t n, const cudaStream_t stream)
+{
+    _get_n_max_index<<<1, 1, 0, stream>>>(labels_sizes_d, nb_label, labels_max_d, n);
```

Comment:
No parallelisation here ? I see you put 1 block and 1 kernel so what is the usage of the kernel ?